_posts/2018-08-09-gsoc-generic-servo.md

```diff
@@ -0,0 +1,92 @@
+
+The biggest problem that caused most of the troubles and pain is that Rust at the time of doing this project did not support generic consts [RFC](https://github.com/rust-lang/rust/issues/44580). 
+
+After two months of solving errors, we have been finally able to compile full servo and take some measurements about which I talk later in this blog post.
+
+Original GSoC project assignment was to prototype ways of splitting script crate, but since results were quite promising and we had a month of coding left we started to prepare a PR to Servo master. During the prototyping period, we have made a few ugly hacks to speed up development. As a result, we had to modify CodeGen to generate generic Bindings and modify thread_local variables that needed to be generic. 
+
+## How it works
+
+The final idea of separation is based on using the *TypeHolderTrait*. TypeHolderTrait is a trait with type parameters for every WebIDL interface that we want to move outside of the script crate. *TypeHolderTrait* itself is defined in script crate with all type parameter traits. However, it is implemented outside of the script crate so it can provide concrete types. TypeHolder enables us to use constructs like static methods in script crate. Later, we use TypeHolder as type parameter for structs and methods that require access to external types. *TypeHolderTrait* might look like:
+
+```rust
+trait TypeHolderTrait {
+    type DomParser: DomParserTrait<Self>
+}
+
+```
+
+````rust
+trait DOMParserTrait<TH: TypeHolderTrait>: MutDomObject + IDLInterface + MallocSizeOf + JSTraceable + DOMParserMethods<TH> {
+    fn Constructor(window: &Window<TH>) -> Fallible<DomRoot<TH::DOMParser>>;
+}
+
+````
+
+IDLTraits usually have only a few functions.
+
+## Effects on Servo
+
+Modifications in final PR should have only minimal effects on Servo speed. However, the codebase has undergone big surgery 🏥. With about 12000 modified line. The most intrusive change is to use TypeHolder where it is required. It turns out that TypeHolder is needed in a lot of places. Leading cause of such a large number of changes is mainly *GlobalScope* which is used all around the codebase and needs to be generic.
+
+Due to lack of Rust support for generic static variables at many places we had to modify the code to use initialization methods to fill static variables. For example, in Bindings, we have added an *InitTypeHolded* function which replaces the content of mutable static variables with appropriate function calls. This pattern was used with the *build.rs* script and also thread locals.
```

Comment:
- "to fill static variables early during program initialization."
- what does the build.rs reference mean?